### ISSUE TITLE
PTX version of test files incompatible with CUDA <= 6.5

### ISSUE THREAD

**Issue by timholy** [JuliaGPU/CUDAdrv.jl#10]
I'm belatedly getting around to updating my lab's internal code to julia-0.5 and more modern packages (we have a *lot* of custom code...), and encountering several problems due to CUDArt/CUDAdrv. I'll report each one here even if I end up being the one to fix them.

First, if I run the version tagged for julia-0.5 (0.1.1, pointing at 5ab568e18ff324887f05b0c12fbd4783e09a0eb2) and try the tests, I get
```jl
julia> include("runtests.jl")
INFO: Recompiling stale cache file /home/tim/.julia/lib/v0.5/CUDAdrv.ji for module CUDAdrv.
device[0]: Tesla C2075, capability 2.0, total mem = 5375 MB
device[1]: Tesla C2075, capability 2.0, total mem = 5375 MB
device[2]: Tesla C2075, capability 2.0, total mem = 5375 MB
ERROR: LoadError: LoadError: No kernel image available/suitable for GPU (CUDA error #209, ERROR_NO_BINARY_FOR_GPU)
 in macro expansion at /home/tim/.julia/v0.5/CUDAdrv/src/base.jl:66 [inlined]
 in CUDAdrv.CuModuleFile(::String) at /home/tim/.julia/v0.5/CUDAdrv/src/module.jl:26
 in include_from_node1(::String) at ./loading.jl:488 (repeats 2 times)
while loading /home/tim/.julia/v0.5/CUDAdrv/test/core.jl, in expression starting on line 151
while loading /home/tim/.julia/v0.5/CUDAdrv/test/runtests.jl, in expression starting on line 6
```

EDIT: if I use the tagged version of CUDArt.jl, which turns out to be prior to the CUDAdrv.jl split, all CUDArt tests pass.

**Comment by timholy** [comment]
Also, before I dive into this package, it might be good to know your thoughts on julia 0.5 support, which I notice is no longer supported on master. Even though I use 0.6 for my personal programming, I won't migrate the big applications of my lab's code to 0.6 until after it's been out for at least a few weeks.

**Comment by maleadt** [comment]
Sorry for breaking your code. I had tried to keep it compatible with CUDArt, but I'm not sure whether its tests have enough coverage.

Anyway, concerning compatibility with julia 0.5: it probably wouldn't take much to get it working again (I removed it when introducing [this constructor syntax](https://github.com/JuliaGPU/CUDAdrv.jl/blob/a34087c1a208debe17a3f0bd87b41045357578dd/src/array.jl#L35-L36) IIRC). Removing support also meant that released versions of CUDArt would be stuck using CUDAdrv v1.x, at which point the integration between both packages was fairly minimal (I didn't want to be too disruptive by pushing much new code in people's environments).

The test error is probably due to loading an incompatible PTX snippet from `test/ptx` (running on 32-bit perhaps?). You can run with the `TRACE=1` environment variable, while disabling precompilation using `--compilecache=no` (the flag is evaluated using `@static` to avoid overhead) to get to see which API call exactly fails.

**Comment by timholy** [comment]
No problem. One way in which I messed up was to not have any `launch` tests that use a `CudaPitchedArray`. My own internal code does test for that, but that doesn't help you very much now, does it? This is surely not the only example of inadequate test coverage, so I chalk another one up to experience.

Regarding that constructor syntax: that's supported by 0.5. Any other reason you bumped the minimum julia version?

Finally, regarding the error: thanks for the debugging tips, very handy indeed. It turns out my system (Tesla cards, CUDA 6.5) wants version 4.1 rather than version 4.3 PTX files. I was able to get past that error by `nvcc -ptx vadd.cu`. Would it be worth copying CUDArt's `deps/` folder structure and building all the test PTX files on the user's machine? One issue is that `global.ptx` doesn't have a corresponding `.cu` file, so that one is causing me problems. (I could edit the file directly, but...)

**Comment by maleadt** [comment]
I [reintroduced 0.5 compatibility](https://github.com/JuliaGPU/CUDAdrv.jl/commit/700ee90cb882a6d462539322da226345a400b82f) (no idea why I originally removed it), and [opened a PR](https://github.com/JuliaGPU/CUDArt.jl/pull/62) to make CUDArt compatible. There's still some issues though, as I've moved to [finalizer-based resource clean-up](https://github.com/JuliaGPU/CUDAdrv.jl#automatic-memory-management) but that seems to conflict with CUDArt's resource management...

As I have a buildbot set-up for CUDAdrv+CUDAnative, I might have a look at setting one up to monitor 0.5 compatibility (when I have some free time).

Regarding the error, I'd rather not introduce a build step just to get the tests working, because it would require users to have CUDA installed (while now they only need the kernel driver, shipping `libcuda.so`). Maybe we should just downgrade the PTX files?

**Comment by timholy** [comment]
Yay!

> I'd rather not introduce a build step just to get the tests working, because it would require users to have CUDA installed (while now they only need the kernel driver, shipping libcuda.so)

Oh, very interesting---that's a great point and very desirable.

As an alternative to downgrading, one could force users to use a certain minimum CUDA/driver version, and just make sure there are tests with helpful error messages. But if you don't need newer features then to me it seems better to use the oldest compatible version.

**Comment by timholy** [comment]
I have a plan, one that is motivated mostly by lack of time. For the short term, I pushed a `standalone` branch to CUDArt that incorporates some of the changes you made but doesn't bring in CUDAdrv. I propose we do one more tag on that branch (it's been ages since I tagged CUDArt) so that we have decent julia-0.5 support for that package.

For the longer term, I'm enthusiastic about relying on CUDAdrv in every way possible: you are the expert, and you're moving whole mountains, when it comes to CUDA and julia. I'll get a lot of benefit from following your lead, once I have time to dig into the breakages and fix them.
